### ISSUE TITLE
Errors resolving dependencies under .Net Core

### ISSUE THREAD

**Issue by SeanSnyders** [tjanczuk/edge#544]
Edge seems to not be dealing with dependencies properly using the `.deps` file under .Net Core.
This has been noticed by @agracio as well as @seansnyders during fixing of general .Net Core compatibility issues as part of a conversation on a commit here:
https://github.com/SeanSnyders/edge/commit/36c58ce807bde561d586141f8a4e3883edfffd6c#commitcomment-21811040

I'll be adding that conversation here as part of this issue such that we can track it properly.

**Comment by SeanSnyders** [comment]
Hi @agracio,

I see you are making changes (and reverting them) to compile for .Net Core. What version of .Net Core are you targeting? I am also in the process of getting Edge working with .Net Core 1.0.0-preview2-1-003177 targeting dependencies `NetCoreApp1.1.0` and `NetStandard1.6`.

Maybe we can work together to get this fixed and working? I'd be up for having a Skype as well.
I am having problems using Edge and .Net Core (the version above) working (with Electron).

Can you please test commit https://github.com/SeanSnyders/edge/commit/439faf8e66ff6cf5331de4fb6dde5feb62542f2c

I get the error below when I test a simple example from [here](https://github.com/SeanSnyders/TestExamples/tree/master/Test-Electron-Edge-DotNetCore):

```
EdgeCompiler::TryCompile (CLR) - Compilation failed with the following errors: 
CS8021: No value for RuntimeMetadataVersion found. No assembly containing System.Object was found nor was a value for RuntimeMetadataVersion specified through options.
```


So far I have not gotten Electron, Edge and .Net Core working. Using the exact same code with .Net Framework works fine...
To debug, I've been hacking around in the dependency manifest reading and loading code of Edge for the internal CLR compiler...

**Comment by SeanSnyders** [comment]
from @agracio:

Hi,
I am using this release version of .Net Core on all platforms: https://www.microsoft.com/net/core#macos. You can see on my branch that all project.json definitions were converted to .csproj as this version of .Net Core no longer supports project.json. Also note that bootstrap.csproj and test.csproj reference Edge.Js.csproj directly rather than referencing nuget package. 
Unlike the errors you getting I have no problems building and partially running on .Net Core, you can see the results here: https://travis-ci.org/agracio/edge/jobs/223456502. Also take a look at travis.yml to see how it runs both `mono` and `.Net Core` builds https://github.com/agracio/edge/blob/master/.travis.yml.
Everything that uses pre-compiled dll works, but anything with C# code fails as it cannot load Edge.Js.CSharp library.
I think this is related to conditional code in `node_modules/edge-cs/lib/edge-cs.js` it executes when it runs in EDGE_USE_CORECLR=1 mode and the fact that `edge-cs` package uses project.json that is not recognised by the version of `.Net Core` I am using. 

I did not have enough time to fully investigate it yet as I was focusing on `StackOverflowException`, will do it later today and let you know. If you have an idea how to resolve Edge.Js.CSharp problem let me know.

My initial goal is to resolve issues on `edge` package (even if its forked as `edge-js`) and later on try to move it to `electron-edge`, just hope that PRs would be accepted on that repo or will have to fork that as well :)

**Comment by SeanSnyders** [comment]
Great thanks for the info.
I have forked your edge repo and are getting the same issues there using `.Net Core 1.0.0-preview2-1-003177`. So it seems the .Net Core version is the problem here - I'll try it again after installing the same version you have.
Would you mind running `dotnet --version` on OSX and posting your result here?

**Comment by SeanSnyders** [comment]
from @agracio

`dotnet --version` returns 1.0.1 on MacOS. 
The problem appears to be with dependencies and the way `edge` is trying to 'compile' them.
If you change `node_modules/edge-cs/lib/edge-cs.js` line 9 to point to `return path.join(__dirname,'../../../', 'lib', 'bootstrap', 'bin', 'Release', 'netcoreapp1.0', 'bootstrap.deps.json');` you will start getting different errors.
This is related to dependency loader https://github.com/tjanczuk/edge/blob/master/src/double/Edge.js/dotnetcore/coreclrembedding.cs#L251 and subsequent invocation in `node_modules/edge-cs/src/Edge.js.CSharp/EdgeCompiler.cs` method `TryCompile`. Some of the libraries added to the dependency resolver return incorrect paths and directory exceptions are raised.
Unfortunately i havent gotten any further as its just to much pain at this point.

**Comment by SeanSnyders** [comment]
Yes! I have also been debugging and seeing errors in the way Edge is loading the dependencies from the manifest `.deps` file. It seems that Edge is not handling dependency packages correctly, so I had to add in this 'fix' on https://github.com/tjanczuk/edge/blob/master/src/double/Edge.js/dotnetcore/coreclrembedding.cs#L296 to traverse the dependencies within a 'package':
![image](https://cloud.githubusercontent.com/assets/1384135/25252223/45efc1a2-2670-11e7-9961-caab7c2a5818.png)

As well as on line https://github.com/tjanczuk/edge/blob/master/src/double/Edge.js/dotnetcore/coreclrembedding.cs#L462 :

![image](https://cloud.githubusercontent.com/assets/1384135/25252320/a1872532-2670-11e7-97d5-c1c1d18ea764.png)

But this is not a total fix just yet.... Still just adding code along the way for debugging the problem.

I think we need to create a new issue and add all this comms to that issue. Agreed?

**Comment by SeanSnyders** [comment]
from @agracio

I tried applying these changes but it made no difference. Unfortunately did not have much time to work on it, hope to make some progress next week.

**Comment by SeanSnyders** [comment]
@agracio the above noted changes does not solve much yet, but Edge currently only looks at the first dependency in the package and ignores the rest, so something like this in the `.deps` file
```
     "runtime.osx.10.10-x64.Microsoft.NETCore.Runtime.CoreCLR/1.1.0": {
        "runtime": {
          "runtimes/osx.10.10-x64/lib/netstandard1.0/SOS.NETCore.dll": {},
          "runtimes/osx.10.10-x64/lib/netstandard1.0/System.Private.CoreLib.dll": {},
          "runtimes/osx.10.10-x64/lib/netstandard1.0/mscorlib.dll": {}
        },
        "native": {
          "runtimes/osx.10.10-x64/native/System.Globalization.Native.dylib": {},
          "runtimes/osx.10.10-x64/native/System.Private.CoreLib.ni.dll": {},
          "runtimes/osx.10.10-x64/native/libcoreclr.dylib": {},
          "runtimes/osx.10.10-x64/native/libdbgshim.dylib": {},
          "runtimes/osx.10.10-x64/native/libmscordaccore.dylib": {},
          "runtimes/osx.10.10-x64/native/libmscordbi.dylib": {},
          "runtimes/osx.10.10-x64/native/libsos.dylib": {},
          "runtimes/osx.10.10-x64/native/mscorlib.ni.dll": {},
          "runtimes/osx.10.10-x64/native/sosdocsunix.txt": {}
        }
```

it only adds `SOS.NETCore.dll` to the runtime and compile dependencies and not any of the others in the `runtime` list. Notably not `mscorlib.dll` which is part of the reason I think I am getting the `System.Object` not found errors.

**Comment by me0262** [comment]
I do want to weigh in on this as well. I'm going through DLL hell trying to get my project running on Linux using SDK 1.0.0-preview2-003131 (CentOS 7) and 1.0.1 runtimes. It's running fine on Windows with the same version.
I'm getting these errors:
CS8021: No value for RuntimeMetadataVersion found. No assembly containing System.Object was found nor was a value for RuntimeMetadataVersion specified through options.
CS0518: Predefined type 'System.Object' is not defined or imported
CS0305: Using the generic type 'Func<T1, T2, T3, T4, T5, T6, T7, T8, T9, TResult>' requires 10 type arguments

I also use external DLLs for running some code, I've had to constantly play with the references in order to get them to work.
Currently, the references are:
"dependencies": {
    "Microsoft.CSharp": "4.3.0",
    "NETStandard.Library": "1.6.1",
    "System.Dynamic.Runtime": "4.3.0",
    "System.Runtime": "4.3.0"
  },
I just recently had to specify System.Dynamic.Runtime, and System.Runtime. These worked on windows without being specified.

I do want to mention that I'm using Electron 1.6.6, with Electron-Edge 6.5.5 (on linux it needs to be rebuilt), on Node.js 7.4.0.

**Comment by agracio** [comment]
I have created a fork in order to provide support for node 7.x and fix some issues. When running `npm test` command locally or running tests on Travis it is able to process external dll calls on `.Net Core`, but fails with the above errors when using `edge-cs`. Can you give it a try and let me know whether you getting the same errors?
- Fork: https://github.com/agracio/edge
- npm package:  `edge-js`
- partially working Travis build log for reference: https://travis-ci.org/agracio/edge/jobs/226125835

**Comment by me0262** [comment]
In my package.json I moved back down to electron 1.4.x / node 6.x.x / edge 6.5.1. (again, success on windows, errors on linux)
I will give this a shot on node 7 when I'm back in the office on Monday, if no one else has tried it then.

**Comment by SeanSnyders** [comment]
@agracio I can repro the installation and testing procedures you mentioned. See https://github.com/agracio/edge/issues/1.

Bar the question how the `EdgeJs.dll` is compiled for distribution via NuGet (when I want to test my own compilation of Edge), I will try and use your `edge-js` npm package on my simple test case https://github.com/SeanSnyders/TestExamples/tree/master/Test-Electron-Edge-DotNetCore and report back.
In that test case, I know I use a [wrapper](https://github.com/SeanSnyders/TestExamples/blob/master/Test-Electron-Edge-DotNetCore/app/mynetcore_edge.cs) c# method to call my function in a pre-compiled c# assembly so that is why I might get errors using the current Edge that seems to fail on c# JIT compilation using .Net Core.
I can try and call directly into the c# assembly and see if that works. If it does, then we are on the same page and it seems only this issue still needs to be addressed to make full .Net Core support work.

@me0262 I hope we can all get on the same page to what works and what not and document that properly. There is a lot of confusion out there - I myself had to painfully plough through info to get to the current state of affairs.

**Comment by SeanSnyders** [comment]
@agracio I can confirm that I can get your fork working using .Net Core on my test case _*if*_ I put the `Task<object> ...` wrapper methods / glue code that interfaces Edge with general c# code inside a pre-compiled assembly.
This is not the general use-case (for me at least), as I put my Edge glue code inside `*.cs` files which are hosted with the Edge/Node project and JIT compiled as part of Edge because the '3rd party' c# code typically comes from other repos/projects. But I can live with this in the mean time and add an extra c# project dependency to my Node projects where the Edge glue code is then pre-compiled.

Can you shed some light on `EdgeJs.dll`? I see that the tests [include that project directly](https://github.com/agracio/edge/blob/master/test/test.csproj#L18) from the repo, but running a test case repo using the npm package `edge-js 6.6.2` with a included .Net Framework or Core project, one still needs a NuGet package for `Edge.js` that comes from your fork that was built from the https://github.com/agracio/edge/tree/master/src/double/Edge.js project.

Is this right? If so, then we are running a mismatch between the compiled node project and the c# part coming from the NuGet package in our test case .Net project that is called from Node.

**Comment by agracio** [comment]
Not sure if I am following you on `EdgeJs.dll` issue, perhaps I am missing something here. I do not see any references to EdgeJs.dll that comes from NuGet. There are some references in `double_stress.csproj` and `double_test.csproj`, but I do not think these play any role in npm package.

**Comment by SeanSnyders** [comment]
The `EdgeJs.dll` NuGet package is compiled as suggested in the docs https://github.com/tjanczuk/edge#building-edgejs-nuget-package

The docs (in https://github.com/tjanczuk/edge#how-to-specify-additional-clr-assembly-references-in-c-code) also suggest to include the NuGet package:

> If you are using .NET Core and are using the .NET Core SDK and CLI, you must have a project.json file (specification here) that specifies the dependencies for the application. This list of dependencies must also include the Edge.js runtime package and, if you need to be able to dynamically compile your code, the package(s) for the compilers that you plan to use, like Edge.js.CSharp.

Your compiled .Net library will produce the `.deps` file that should be used by Edge to load the required dependencies of your .Net code to execute properly.

So the Edge NuGet package that includes `EdgeJs.dll` is a companion to the Edge npm package.

**Comment by agracio** [comment]
Sorry, looks like I am really slow trying to catch up with this issue. 
So original `edge` repo indeed uses EdgeJs Nuget package for .Net Core. However since I cannot control that package I switched it to use .csproj reference directly, and that way it is compiled on `npm install`.
As you have noticed test project is referencing .csproj directly, but more importantly so does `lib/bootstrap/bootstrap.csproj`, this will make it recompile EdgeJs and use it as a reference in `.deps` file. 
I think I am completely missing the importance of Nuget EdgeJs package here, as to me it appears that it can be replaced with direct reference to project.
You have a considerably better understanding of edge inner workings on .Net Core , so you would need to explain it in to me, perhaps a Skype chat would help :)

**Comment by agracio** [comment]
Had some time to work on .Net Core problems today and was able to resolve/hack a couple of issues. 
- On my version of .Net Core on MacOS (1.0.1) some of compile dependencies were only added as lower case, so for example `system.runtime` would be lower cased only, and since edge compile dependency resolution uses dictionaries, it would fail to resolve `System.Runtime`. Not sure why that happens on my mac and doesn't happen on yours though. Forcing names to be upper case solves the issue.
- Some dependencies added to dictionary are pointing to the non existing .dll path. Even if correct path exists it will not be added as the key already exists in the dictionary. Adding a simple File.Exists() check resolves that too.

Once the above issues are resolved I am able to run both dlls and edge-cs code. 

One thing I've noticed is that assembly resolution happens before every edge-cs func execution and it adds rather significant overhead.

Let me know if you had any progress on your side.

**Comment by me0262** [comment]
Hmm... I'm still getting the same.
Windows no problem.
Linux CentOS 7 / Node 7.4.0 / Electron 1.6.6 / NET Core 1.0.0-preview2-3131 still gets error.

I've also attempted to put System.Runtime.dll into the project, but comes up with the same errors.
I'm going to try and see how this works without DLLs or references...

----> Errors when compiling as a CLR async lambda expression:
CS1701: Assuming assembly reference 'System.Runtime, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' used by 'System.Threading.Tasks' matches identity 'System.Runtime, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' of 'System.Runtime', you may need to supply runtime policy
CS1701: Assuming assembly reference 'System.Runtime, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' used by 'System.Threading.Tasks' matches identity 'System.Runtime, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' of 'System.Runtime', you may need to supply runtime policy
CS0012: The type 'Object' is defined in an assembly that is not referenced. You must add a reference to assembly 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'.

**Comment by agracio** [comment]
@me0262 Currently this is in dev stages and is not available on electron-edge. Once there is a solution we will submit a PR to electron-edge repo.
You can try dev branch here: https://github.com/agracio/edge/tree/net-core, but its only for node, not Electron.

**Comment by SeanSnyders** [comment]
@agracio I've unfortunately been busy trying to solve problems that circumvents the use of Edge to make progress on those projects that was being held up by Edge not working on .Net Core.

But what you explored sounds like a good direction. I will check the caps issue with my install and use your new branch to test.
I will also test it with Electron as this is where I get the deeper dependency issues. And here I think that the https://github.com/tjanczuk/edge/issues/544#issuecomment-296577774 I made might still apply....

@me0262 @agracio You can still test it with Electron.
The Electron-Edge project is only a 'easy' way to make sure that the `NODE_MODULE_VERSION` for node modules with native dependencies align between Electron and Edge, and is just a recompile of Edge to achieve that. Because we are recompiling Edge anyways, you should be able to use this dev branch to test with Electron.
The way I achieved that (because Electron 1.6.5/6 uses `NODE_MODULE_VERSION` 53 and the latest Node 7.x is only on 51), is to put a `.npmrc` file in the Edge folder that will be recompiled.
This file contains this:
```
runtime = electron
target = 1.6.5
target_arch = x64
disturl = https://atom.io/download/atom-shell
```
When Edge is recompiled, it will target the same node module version as the Electron specified in the `.npmrc` file.
After the recompile, then set the Edge variable `EDGE_NATIVE` to the fully qualified location of `edge_coreclr.node` file that the compilation made.

So when your app then runs, it will use that Edge. You can still specify the `edge-js` dependency in your `package.json`.

I hope that makes sense....

**Comment by agracio** [comment]
Managed to get Travis OSX build working, and it appears that with my fixes the build is passing not only on my Mac but also on Travis build workers: https://travis-ci.org/agracio/edge/jobs/227707180 except for a couple of failing tests that need to be addressed.
Linux is still failing, will have a look at it in the next couple of days.
@SeanSnyders by the looks of your tests output you already had everything working, have you tried running any edge-cs code?
Note the change I've made to manifest loading btw: https://github.com/agracio/edge/blob/net-core/lib/edge.js#L119.

**Comment by me0262** [comment]
@SeanSnyders I've worked around that issue as well. I've installed python 2.7 onto my machine, and am using electron-rebuild to recompile any modules (it hooks immediately to whatever edge I've been working with), after that, the version error goes away.

I know what you mean, I've been teating my hair out for the past 2 weeks trying to get this running on Linux. It's a major dependency for the client.

I have also been able to switch to whatever edge project I've wanted. My C# code is placed on the main process and not being called from the HTML renderer.

I've worked around my errors with System.Runtime and System.Assembly as well, I was loading unreferenced DLLs at runtime. After I took those out (the project no longer calls for them anyways), I then got the error where it could not find the referenced DLL (again, working fine on Windows). 
After specifying EDGE_USE_CORECLR=1 and EDGE_APP_ROOT=app/lib/ (path to my dlls and deps.json), I now get 
CoreCLREmbedding::Initialize - Failed to initialize CoreCLR, HRESULT 0x80070057
TypeError: edge.initializeClrFunc is not a function.

I'm trying the EDGE_NATIVE=path/ that you mentioned next...

@agracio looks like in 6.6.3 you updated the project file to the new *.csproj format, and removed the project.json. (Compiling bootstrap for Net Core App 1.0)
edge-cs (dependency) still uses the old project.json files.
I was running 1.0.0-preview2_3131 so I would be able to maintain compatibility.
I'll switch to my 1.0.1 VM snapshot and try there.

**Comment by agracio** [comment]
@me0262 I left project.json files in the previous release by mistake, only working to support more recent versions of .Net Core versions.
But the good news is that I've now managed to resolve both MacOS and Linux .net Core builds, at least as far as running tests is concerned. It was all due to lower case assembly naming and the fix is trivial https://github.com/agracio/edge/blob/net-core/src/double/Edge.js/dotnetcore/coreclrembedding.cs#L182.
Take a look at this Travis build https://travis-ci.org/agracio/edge/builds/228132650, both Linux and MacOS builds are now running tests on .Net Core. The failing tests are not something that would ever cause problems in real life applications.
This is still running on a branch https://github.com/agracio/edge/tree/net-core, but I am now considering merging it back to master. @SeanSnyders, @me0262 could you verify whether you can run .Net Core code on that branch?

**Comment by SeanSnyders** [comment]
@agracio sounds great. I'll do some testing, but unfortunately I can only do that with `project.json` support. Would it be possible to keep these there and updated until we have a stable .Net Core Edge solution? Then one can migrate away in a systematic way and by correct semver changes in Edge's versioning...?

**Comment by agracio** [comment]
I have merged my .Net Core fixes to the main branch and released it as 6.6.4, it also includes project.json files. 
But I have not tested it using project.json files, so somewhat sceptical it will work. Everything seams to work fine on .Net Core 1.0.1 and above though.

**Comment by SeanSnyders** [comment]
Okay thanks, I'll try and test 6.6.4.
